CDS-58583 config files topic (#1402)
harness/developer-hub: docs/continuous-delivery/x-platform-cd-features/services/cd-services-config-files.md

---
title: Use config files in your deployments
description: Use plain text and encrypted files in your deployments
sidebar_position: 6
---

You can use files added to the **Config Files** section in your Harness services in your manifests, such as in a ConfigMap. You can reference unencrypted and encrypted files, and they can be single or multiline.

## Supported platforms

All platform [integrations](/docs/continuous-delivery/integrations/cd-integrations) (Kubernetes, etc) support config files stored in the Harness File Store. Eventually, Harness will support config files from Git platforms.

## Limitations

- Files must be 1MB or less.

> - All file types are supported.

Reviewer: This doesn't seem to fit under "limitation". I think you could stick it at the end of the first paragraph (line 7).

Author: changed heading. thanks

- You cannot use Harness variables in an encrypted text config file.
- You cannot reference other config files within a config file.

## Config file capabilities

Harness supports the ability to add any file type to a service. Common types like JSON and XML are popular in our SSH and WinRM integrations. 

You can add plain or encrypted text files. Both types can be referenced using a Harness expression, discussed below.

With a plain text config file, Harness renders the contents of the file. 

> With an encrypted text config file, you need to base64 decode it before you can reference it within the deployment. This is described below.

Reviewer: could you use anchored links to the sections instead of saying "discussed/described below"?

You can attach multiple files in a config file. Simply add a new line:

```yaml
                  files:
                    - /Config files/file1.json
                    - /dev/file2.json
```

```mdx-code-block
  </TabItem>
  <TabItem value="Pipeline Studio" label="Pipeline Studio">
```

1. In the Harness service, in Config Files, select **Add Config File**.
2. In **Config File Source**, select **Harness**, and select **Continue**.
3. In **Config File Identifier**, enter a name for the file.
4. In **Select file type**, select **File Store** or **Encrypted**. Encrypted files are stored as [Harness secrets](/docs/platform/secrets/secrets-management/harness-secret-manager-overview/).
5. Select **Add** to attach multiple files as a single config file.
6. Select **Submit**.

You will use the value you entered in **Config File Identifier** to reference the config file.

```mdx-code-block
  </TabItem>
</Tabs>
```

## Multiple files can be added to one config file

You can attach multiple files to one config file. All the files must be either plain text or encoded. You cannot mix types.

## Referencing and encoding config files

> Files added in the **Config Files** section of a service are referenced using the `<+configFile.getAsString("config_file_Id")>` Harness expression:
> 
> * `<+configFile.getAsString("config_file_Id")>` - Plain text file contents.
> * `<+configFile.getAsBase64("config_file_Id")>` - Base64-encoded file contents.

Reviewer: ```suggestion
Files added in the **Config Files** section of a service are referenced using the Harness expression `<+configFile.getAsString("config_file_Id")>`.

* Plain text file contents: `<+configFile.getAsString("config_file_Id")>`
* Base64-encoded file contents: `<+configFile.getAsBase64("config_file_Id")>`
```

For me, it is a bit easier to read with the code samples last.

### Use Base64 to avoid new lines

If you are going to use a config file in a manifest or shell script, be aware that `<+configFile.getAsString()>` can cause problems by adding new lines to your manifest (unless you have formatted the file very carefully).

> Instead, use `<+configFile.getAsBase64()>`. This will ensure that the contents of the file are rendered as a single line.

Reviewer: ```suggestion
Instead, use `<+configFile.getAsBase64()>`. This ensures that the contents of the file are rendered as a single line.
```

6. Follow the steps to select the override file, and then select **Submit**. 

When the selected service is deployed to this environment, the environment's config files will override that service's config files.

### Fully overriding config files and variables

Harness supports [overriding service](/docs/continuous-delivery/x-platform-cd-features/environments/create-environments) config files at the environment level. 

Config files are a black box that can contain multiple formats and content, such as YAML, JSON, plain text, etc. Consequently, they cannot be overridden like Values YAML files.

When you have config files at two or more of the environment Service Overrides, Configuration, and the service itself, the standard override priority is applied.

The priority from top to bottom is:

1. Environment service overrides
2. Environment configuration
3. Service settings

![picture 1](static/9fbf54319fcbbfaf81732ede00d0e6226ea193524fd8fe61dae5ade93ec6fc71.png)  

When you have Variables with the same name at two or more of the environment Service Overrides, Configuration, and the service itself, the standard override priority is applied.

> ## Use Copy Configs Command

Reviewer: ```suggestion
## Use the Copy Configs command
```
